### PR TITLE
Add option to cache user for access token

### PR DESCRIPTION
**Pull request by joehoyle**
This can significantly reduce the load on the origin delegated server

### CODE REVIEW

File: inc/oauth2/namespace.php [hunk 1]
```diff
@@ -118,10 +119,25 @@ function attempt_authentication( $user = null ) {
...
+			$local_user = new WP_User( $cached_id );
+		} else {
+			$local_user = synchronize_user_for_token( $token_value );
+			if ( is_wp_error( $local_user ) ) {
+				$delegated_auth_error = $local_user;
+				return $user;
+			}
+			wp_cache_set( $token_value, $local_user->ID, 'local_user_id_for_token', HM_DELEGATED_AUTH_ACCESS_TOKEN_CACHE_TTL );
+		}
+	} else {
```
Comment (rmccue) [inline inc/oauth2/namespace.php]: Can we return early to avoid this if tree instead?

File: inc/oauth2/namespace.php [hunk 2]
```diff
@@ -118,10 +119,25 @@ function attempt_authentication( $user = null ) {
 
 	$is_querying_token = true;
 
-	$local_user = synchronize_user_for_token( $token_value );
-	if ( is_wp_error( $local_user ) ) {
-		$delegated_auth_error = $local_user;
-		return $user;
+	// Locally cache the user for the given access token if it is enabled.
+	if ( defined( 'HM_DELEGATED_AUTH_ACCESS_TOKEN_CACHE_TTL' ) ) {
+		$cached_id = wp_cache_get( $token_value, 'local_user_id_for_token' );
```
Comment (rmccue) [inline inc/oauth2/namespace.php]: Do we want to define this as a global group?

**Review by rmccue — CHANGES_REQUESTED** [review]
Also, needs docs.

File: inc/oauth2/namespace.php (hunk 2 above)
Comment (joehoyle) [reply inc/oauth2/namespace.php]: yeah I think so

File: inc/oauth2/namespace.php (hunk 1 above)
Comment (joehoyle) [reply inc/oauth2/namespace.php]: Done!

File: inc/oauth2/namespace.php (hunk 2 above)
Comment (joehoyle) [reply inc/oauth2/namespace.php]: Added in a new `bootstrap()` function.

File: README.md [hunk 3]
```diff
@@ -9,3 +9,11 @@ When a login request is delegated to the upstream WordPress site a local cache o
 For Cookie auth to work correctly you must create an OAuth 2 application on the upstream WordPress site. Do so using the `home_url( '/hm-delegated-auth-callback' )` as the Callback URL, and configure this plugin with the Client ID in the PHP constant `HM_DELEGATED_AUTH_CLIENT_ID`.
 
 Note that if you are using multisite, this must use the home URL for the main site so that the callback URL is the same for the whole network. Delegated Auth then redirects internally between sites as necessary.
+
+## Caching Access Token Authentication
+
+Check the upstream WordPress site on each HTTP request to validate the OAuth 2 access token can cause significant load. Delegated Auth support locally caching the access token validation for a short period of time, so not every request causes an upstream HTTP request. To enable this functionality, define the `HM_DELEGATED_AUTH_ACCESS_TOKEN_CACHE_TTL` constant with an integer value in seconds. For example, to cache Access Token validation for 60 seconds:
```
Comment (rmccue) [inline README.md]:
```suggestion
Checking the upstream WordPress site on each HTTP request to validate the OAuth 2 access token can cause significant load. Delegated Auth supports locally caching the access token validation for a short period of time, so not every request causes an upstream HTTP request. To enable this functionality, define the `HM_DELEGATED_AUTH_ACCESS_TOKEN_CACHE_TTL` constant with an integer value in seconds. For example, to cache Access Token validation for 60 seconds:
```

File: inc/oauth2/namespace.php [hunk 4]
```diff
@@ -5,6 +5,18 @@
 use function HM\Delegated_Auth\synchronize_user_for_token;
 use WP_Error;
 use WP_Http;
+use WP_User;
+
+const ACCESS_TOKEN_CACHE_GROUP = 'cache-users-for-access-token';
+
+/**
+ * Bootstrapp the OAuth 2 adaptor.
```
Comment (rmccue) [inline inc/oauth2/namespace.php]:
```suggestion
 * Bootstrap the OAuth 2 adaptor.
```

**Review by rmccue — CHANGES_REQUESTED** [review]

**Review by rmccue — APPROVED** [review]

**Pull request merged**